### PR TITLE
Archive additional metadata and other information

### PR DESCRIPTION
**Pull request by andrew-c-ross**
@charliestock and I have been talking with the data portal team about metadata to include in the latest files that will be served on the CEFI data portal. We largely settled on adding an attribute to the published netcdf files pointing to the path of the data on /archive, and then storing additional metadata and other information in /archive. This PR adds some of that extra metadata to a new `metadata.out/` directory that gets included in the ascii tar file. The main additions here are archiving the model XML file itself and the commit hash for each git submodule in the source code.

The new `git_submodule_status` file that gets archived will look something like this; it is the result of running `git submodule status --recursive`:
```
 f703b82972701a4e32a46d4d44a15f9fc2debb27 src/FMS (2024.03)
 8f96707a90132ca119d81ed84e5a62ca0ff3ed96 src/Icepack (Icepack1.1.0-139-g8f96707)
 3fb64c40e21d453fad18ad19de69c6c30a846a64 src/MOM6 (remotes/origin/dev_cefi_backup-62-g3fb64c40e)
 9423197f894112edfcb1502245f7d7b873d551f9 src/MOM6/pkg/CVMix-src (9423197)
 29e64d652786e1d076a05128c920f394202bfe10 src/MOM6/pkg/GSW-Fortran (29e64d6)
 022bd89d1bfe03dc259afc24a6e19c334e47eb4b src/SIS2 (fix_esm_dust_flux_2016.05.26-920-g022bd89)
 ff2cce7820bbe4d23b7cc82d15ed57f80fc7a57d src/atmos_null (xanadu_esm4_20190304-10-gff2cce7)
 e84cf65f1622c31f3143a793c11378f050398907 src/coupler (2024.03)
 1d19a57e0be2ba1b668e2f0dd7a7290733b21cfe src/ice_param (2020.01-alpha1-19-g1d19a57)
 fd6478fecd8891aebcba904fe4b34a415fe30e73 src/icebergs (remotes/origin/dev/master-134-gfd6478f)
 58cee106df801910d3372321cf8bb105e362aa48 src/land_null (2020.01-alpha1-4-g58cee10)
 840b65c40675e2d06bf40405ad3f12dec7f35923 src/libyaml (0.2.5-8-g840b65c)
 43655806c942ebb8a0750d5c70bb5259e658ae1a src/mkmf (2023.01-8-g4365580)
 0de1ab230a12999eb139763f40e56920a204d639 src/ocean_BGC (2022.03-78-g0de1ab2)
 385222c469942f0562b4c70b926dfdd8173138e7 src/ocean_BGC/mocsy (v2.3.3-6-g385222c)
```

This PR could also be a good place to discuss if there are other files that should be added to the archive.

### CODE REVIEW

**Comment by andrew-c-ross** [comment]
Also, let me know if this should go here or in the regional-mom6-xml repo

**Comment by yichengt900** [comment]
> Also, let me know if this should go here or in the regional-mom6-xml repo

@andrew-c-ross, thank you for adding the extra metadata to the archive. I will review it more closely later, but at first glance, I didn’t notice any major issues. I believe it would be beneficial to apply these changes to our internal repository as well, as it would allow other developing regions to benefit from the extra metadata in the archives.

**Comment by chiaweh2** [comment]
Thanks @andrew-c-ross @yichengt900! I will have a global attribute of `cefi_archive_version` that provide the absolution path in the /archive for every single netcdf files.

File: xmls/NEP10/CEFI_NEP_cobalt.xml [hunk 1]
```diff
@@ -690,15 +690,23 @@ COBALT_INPUT_EOF
...
+         mkdir extra.results
+         mv seaice.stats* ocean.stats* *velocity_truncations extra.results/
+         # Also archive an assortment of model input files and model metadata
+         mkdir metadata.out
+         mv MOM_parameter_doc* SIS_parameter_doc* COBALT_parameter_doc* metadata.out/
+         cp diag_table input.nml metadata.out/
+         cp INPUT/*_input INPUT/*_override metadata.out/
+         # Save the Git commit hash for each source code submodule
+         git -C $root/$(LIBS_ROOT)/src/mom6 submodule status --recursive > metadata.out/git_submodule_status
+         # Archive a copy of the xml used to run the model and the run scipt (fre sets these)
```
Comment (yichengt900) [inline xmls/NEP10/CEFI_NEP_cobalt.xml:703]: `scripts`

File: xmls/NEP10/CEFI_NEP_cobalt.xml [hunk 2]
```diff
@@ -690,15 +690,23 @@ COBALT_INPUT_EOF
...
+         cp $rtsxml metadata.out/
+         cp $scriptName metadata.out/runscript.csh
+         # When the ocean uses a mask_table the ocean_static.nc file produced by the model run has holes in coordinates (geolon,geolat)
+         # This causes problems for ferret and python tools for analysis.
+         # Copy a non-masked version of ocean_static.nc to be saved as a history file to be used by the analysis scipts.
```
Comment (yichengt900) [inline xmls/NEP10/CEFI_NEP_cobalt.xml:708]: `scripts`

File: xmls/NWA12/CEFI_NWA12_cobalt.xml [hunk 3]
```diff
@@ -692,13 +692,22 @@ COBALT_INPUT_EOF
...
-         #Copy a non-masked version of ocean_static.nc to be saved as a history file to be used by the analysis scipts .
+         mv seaice.stats* ocean.stats* *velocity_truncations extra.results/
+         # Also archive an assortment of model input files and model metadata
+         mkdir metadata.out
+         mv MOM_parameter_doc* SIS_parameter_doc* COBALT_parameter_doc* metadata.out/
+         cp diag_table input.nml metadata.out/
+         cp INPUT/*_input INPUT/*_override metadata.out/
+         # Save the Git commit hash for each source code submodule
+         git -C $root/$(LIBS_ROOT)/src/mom6 submodule status --recursive > metadata.out/git_submodule_status
+         # Archive a copy of the xml used to run the model and the run scipt (fre sets these)
```
Comment (yichengt900) [inline xmls/NWA12/CEFI_NWA12_cobalt.xml:705]: `script`

File: xmls/NWA12/CEFI_NWA12_cobalt.xml [hunk 4]
```diff
@@ -692,13 +692,22 @@ COBALT_INPUT_EOF
...
+         cp $rtsxml metadata.out/
+         cp $scriptName metadata.out/runscript.csh
+         # When the ocean uses a mask_table the ocean_static.nc file produced by the model run has holes in coordinates (geolon,geolat)
+         # This causes problems for ferret and python tools for analysis.
+         # Copy a non-masked version of ocean_static.nc to be saved as a history file to be used by the analysis scipts.
```
Comment (yichengt900) [inline xmls/NWA12/CEFI_NWA12_cobalt.xml:710]: `scripts`

File: xmls/NWA12/CEFI_NWA12_cobalt_fms2_yaml.xml [hunk 5]
```diff
@@ -695,13 +695,22 @@ diag-table-to-yaml -f $work/diag_table -o $work/diag_table.yaml
...
-         #Copy a non-masked version of ocean_static.nc to be saved as a history file to be used by the analysis scipts .
+         mv seaice.stats* ocean.stats* *velocity_truncations extra.results/
+         # Also archive an assortment of model input files and model metadata
+         mkdir metadata.out
+         mv MOM_parameter_doc* SIS_parameter_doc* COBALT_parameter_doc* metadata.out/
+         cp diag_table input.nml metadata.out/
+         cp INPUT/*_input INPUT/*_override metadata.out/
+         # Save the Git commit hash for each source code submodule
+         git -C $root/$(LIBS_ROOT)/src/mom6 submodule status --recursive > metadata.out/git_submodule_status
+         # Archive a copy of the xml used to run the model and the run scipt (fre sets these)
```
Comment (yichengt900) [inline xmls/NWA12/CEFI_NWA12_cobalt_fms2_yaml.xml:708]: `script`

File: xmls/NWA12/CEFI_NWA12_cobalt_fms2_yaml.xml [hunk 6]
```diff
@@ -695,13 +695,22 @@ diag-table-to-yaml -f $work/diag_table -o $work/diag_table.yaml
...
+         cp $rtsxml metadata.out/
+         cp $scriptName metadata.out/runscript.csh
+         # When the ocean uses a mask_table the ocean_static.nc file produced by the model run has holes in coordinates (geolon,geolat)
+         # This causes problems for ferret and python tools for analysis.
+         # Copy a non-masked version of ocean_static.nc to be saved as a history file to be used by the analysis scipts.
```
Comment (yichengt900) [inline xmls/NWA12/CEFI_NWA12_cobalt_fms2_yaml.xml:713]: `scripts`

**Review by yichengt900 — COMMENTED** [review]
@andrew-c-ross, the section on metadata and additional information looks good to me. I think it covers almost everything. I only found a few small typos. I will leave this PR open a little longer so others have time to review and provide any feedback.

**Comment by andrew-c-ross** [comment]
Thanks, I should spellcheck before I copy and paste 🤦

**Comment by theresa-morrison** [comment]
I like this idea. Saving the commit hash for each module seems especially helpful for reproducing old simulations since defaults can change over time. I think moving the parameter doc iles into the meta data folder also makes sense.

Should we use this as an opportunity to rename `extra.results`? I would propose either `stats` or `model.stats` based on what goes into those folders. There should be no velocity truncations, so that file would not be included in `stats` most of the time.

Is the `field_table` worth saving in the meta data? or are they generic enough we don't want to include them?

**Comment by theresa-morrison** [comment]
If the field table is in the xml that should be good enough. 

I would suggest we test if `extra.results` is needed with new versions of FRE, and remove it if it is not.

**Comment by yichengt900** [comment]
Good catch, @andrew-c-ross. I can confirm that the new FRE now willl archive stats files as well as velocity_truncation files (if you name them like U.velocity_truncations).

**Comment by andrew-c-ross** [comment]
> (if you name them like U.velocity_truncations)

I think they are currently named like U_velocity_truncations. Does that not get caught? It looks like the name can be changed with the `U_TRUNC_FILE` and `V_TRUNC_FILE` MOM_input options?

**Comment by yichengt900** [comment]
@andrew-c-ross, yes, we can always change file names in the MOM_input options. The issue is that the new FRE treats "velocity_truncation" as a file extension, so unfortunately, it cannot recognize `U_velocity_truncations` or `V_velocity_truncations`.

**Review by yichengt900 — APPROVED** [review]
@andrew-c-ross, I have tested the new changes and can confirm that they successfully archive the metadata and additional information as proposed. Approved. I'll wait a bit to see if there is any further feedback. Thank you again for this contribution!

**Comment by yichengt900** [comment]
I will merge this PR for now. If any metadata-related issues arise or concerns are raised later, we can always revisit them. We'll also add these changes to our private XML repository later.

**Pull request merged**